assignment/assignment-impl/impl/src/java/org/sakaiproject/assignment/impl/BaseAssignmentService.java

```diff
@@ -6015,9 +6279,68 @@ public void handleAccess(HttpServletRequest req, HttpServletResponse res, Refere
 					Collection copyrightAcceptedRefs) throws EntityPermissionException, EntityNotDefinedException,
 					EntityAccessOverloadException, EntityCopyrightException
 			{
+				M_log.debug("handleAccess from getHttpAccess ");
+				M_log.debug("ref " + ref.getId() + " - " + ref.getSubType() + " - " + ref.getReference() + " - " + ref.getContainer());
 				if (SessionManager.getCurrentSessionUserId() == null)
 				{
-					// fail the request, user not logged in yet.
+					M_log.debug("user not logged in");
+					try{
+						if ("s".equals(ref.getSubType()))
+						{
+							M_log.debug("getting submission");
+							//TODO should we check the assignment settings?
+							Session newsession = SessionManager.startSession();
+							newsession.setActive();
+							SessionManager.setCurrentSession(newsession);
+							newsession.setUserId("admin");
+							newsession.setUserEid("admin");
```

Comment:
Consider pseudocode like:
currentUser = getCurrentUser(); setCurrentUser(admin); do stuff; restoreCurrentUser();
I've encountered this issue before - simply by double clicking, you spawn two threads:
Thread 1:
currentUser = ""; setCurrentUser(admin); do stuff;
Thread 2:
currentUser = admin; setCurrentUser(admin); do stuff;
thread 1:
restoreCurrentUser() // becomes ""
thread 2:
restoreCurrentUser() // becomes "admin"

Then your user remains an administrator, and very very bad things happen. SecurityAdvisor is tied to the current thread and I highly advise using it instead.